unittest/Conversion/TritonGPUToLLVM/DumpLayout.cpp

```diff
@@ -40,15 +50,34 @@ class IndexEmitter {
 public:
   IndexEmitter(MLIRContext *context_)
       : context(context_), option(context), typeConverter(context, option),
-        rewriter(context), loc(UnknownLoc::get(context)) {
-    rewriter.setInsertionPointToStart(&block);
+        rewriter(context), loc(UnknownLoc::get(context))
+#ifdef USE_ROCM
+        ,
+        amdHelper(typeConverter)
+#endif
+  {
+    mlir::OpBuilder builder(context);
+    std::vector<mlir::Type> inTypes{};
+    std::vector<mlir::Type> outTypes{};
+    auto funcTy = builder.getFunctionType(inTypes, outTypes);
+    auto func = builder.create<mlir::triton::FuncOp>(loc, "test_func", funcTy);
+    auto mlirModule = mlir::ModuleOp::create(loc);
+    mlirModule.push_back(func);
+    auto *block = func.addEntryBlock();
+    rewriter.setInsertionPointToStart(block);
   }
 
   llvm::SmallVector<llvm::SmallVector<Value>>
   emitIndices(Attribute layout, llvm::ArrayRef<int64_t> shape,
               bool withCTAOffset) {
     auto type = RankedTensorType::get(shape, rewriter.getF16Type(), layout);
-    return mlir::emitIndices(loc, rewriter, layout, type, withCTAOffset);
+    return
+#ifdef USE_ROCM
+        amdHelper.emitIndices(
+#else
+        emitIndices(
+#endif
```

Comment:
that kind of code is not going to be maintainable. Can you refactor this code to layer things properly, either pass different parameters or use inheritance to be able to control to call different backends.

Reply:
@ThomasRaoux , Is ``#ifdef USE_ROCM`` a red flag for merging? should I get rid of them?

Reply:
So, using ``#ifdef USE_ROCM`` is a definitely a wrong way. I've introduced new directives and build different targets for Nvidia and AMD.
Unfortunately some of the tests are failed, so I disabled them. Please take a look